Avoid clone for sparse tensors during accumulation of grads.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33427 Avoid clone for sparse tensors during accumulation of grads.**
* #33214 Unify gradient accumulation between distributed autograd and local autograd
engine.
engine.
engine.
engine.
engine.
engine.
engine.
engine.
engine.
engine.
engine.
engine.

This PR is an attempt to avoid clone for sparse tensors similar to how
we avoid clone for dense tensors currently.

As per my understanding even if the 'indices' and 'values' of a sparse tensor
are non-continguous, operations like 'add' are still supported. As a result,
the major change in this PR is to use create a shallow copy instead of clone()
for sparse tensors.

Differential Revision: [D19926698](https://our.internmc.facebook.com/intern/diff/D19926698/)

pritamdamania committed:
Avoid clone for sparse tensors during accumulation of grads.

This PR is an attempt to avoid clone for sparse tensors similar to how
we avoid clone for dense tensors currently.

As per my understanding even if the 'indices' and 'values' of a sparse tensor
are non-continguous, operations like 'add' are still supported. As a result,
the major change in this PR is to use create a shallow copy instead of clone()
for sparse tensors.

Differential Revision: [D19926698](https://our.internmc.facebook.com/intern/diff/D19926698/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Avoid clone for sparse tensors during accumulation of grads."

engine.

This PR is an attempt to avoid clone for sparse tensors similar to how
we avoid clone for dense tensors currently.

As per my understanding even if the 'indices' and 'values' of a sparse tensor
are non-continguous, operations like 'add' are still supported. As a result,
the major change in this PR is to use create a shallow copy instead of clone()
for sparse tensors.

Differential Revision: [D19926698](https://our.internmc.facebook.com/intern/diff/D19926698/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Avoid clone for sparse tensors during accumulation of grads."

engine.
engine.

This PR is an attempt to avoid clone for sparse tensors similar to how
we avoid clone for dense tensors currently.

As per my understanding even if the 'indices' and 'values' of a sparse tensor
are non-continguous, operations like 'add' are still supported. As a result,
the major change in this PR is to use create a shallow copy instead of clone()
for sparse tensors.

Differential Revision: [D19926698](https://our.internmc.facebook.com/intern/diff/D19926698/)

[ghstack-poisoned]

albanD commented on test/test_autograd.py:
```diff
@@ -3240,6 +3240,78 @@ def backward(ctx, grad):
+        self.assertTrue(p_a == p_g or p_b == p_g)
+
+        # Run backwards multiple times to ensure accumulation works.
+        for i in range(100):
```
Maybe 10 is enough here? We want to keep the runtime of tests low.

albanD commented on test/test_autograd.py:
```diff
@@ -3240,6 +3240,78 @@ def backward(ctx, grad):
+        for i in range(100):
+            loss.backward(retain_graph=True)
+
+        # non-contiguous indices and value should not trigger copy either.
```
Maybe they should? The current behavior would return contiguous indices and values all the time no?

pritamdamania87 replied:
I'm not sure about this, is this a contract we make with users currently? If not it seems like we should optimize this codepath as well. Are there any ops on SparseTensors that fail if indices and values are non-contiguous? Based on what I saw, we usually handle non-contiguous indices and values in SparseTensors appropriately.

albanD replied:
I'm afraid there is no contract :/ So anything that changes the current behavior is a BC-breaking change.
A user could be giving this directly into a custom cpp function that assumes that indices and values are contiguous for example.

albanD commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
```diff
@@ -1869,6 +1869,82 @@ def backward(ctx, grad):
+            self.assertTrue(p_a == p_g or p_b == p_g)
+
+            # Run backwards multiple times.
+            for i in range(100):
```
Same comment as above for the number of iterations.

albanD reviewed (commented):
As discussed offline, if we want to double check if this is BC breaking or not.
So that we either make it BC or we mark the PR as BC-breaking.

pritamdamania committed:
Update on "Avoid clone for sparse tensors during accumulation of grads."

engine.
engine.
engine.

This PR is an attempt to avoid clone for sparse tensors similar to how
we avoid clone for dense tensors currently.

As per my understanding even if the 'indices' and 'values' of a sparse tensor
are non-continguous, operations like 'add' are still supported. As a result,
the major change in this PR is to use create a shallow copy instead of clone()
for sparse tensors.

Differential Revision: [D19926698](https://our.internmc.facebook.com/intern/diff/D19926698/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Avoid clone for sparse tensors during accumulation of grads."

engine.
engine.
engine.
engine.

This PR is an attempt to avoid clone for sparse tensors similar to how
we avoid clone for dense tensors currently.

As per my understanding even if the 'indices' and 'values' of a sparse tensor
are non-continguous, operations like 'add' are still supported. As a result,
the major change in this PR is to use create a shallow copy instead of clone()
for sparse tensors.

Differential Revision: [D19926698](https://our.internmc.facebook.com/intern/diff/D19926698/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Avoid clone for sparse tensors during accumulation of grads."

engine.
engine.
engine.
engine.
engine.

This PR is an attempt to avoid clone for sparse tensors similar to how
we avoid clone for dense tensors currently.

As per my understanding even if the 'indices' and 'values' of a sparse tensor
are non-continguous, operations like 'add' are still supported. As a result,
the major change in this PR is to use create a shallow copy instead of clone()
for sparse tensors.

Differential Revision: [D19926698](https://our.internmc.facebook.com/intern/diff/D19926698/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Avoid clone for sparse tensors during accumulation of grads."

engine.
engine.
engine.
engine.
engine.
engine.

This PR is an attempt to avoid clone for sparse tensors similar to how
we avoid clone for dense tensors currently.

As per my understanding even if the 'indices' and 'values' of a sparse tensor
are non-continguous, operations like 'add' are still supported. As a result,
the major change in this PR is to use create a shallow copy instead of clone()
for sparse tensors.

Differential Revision: [D19926698](https://our.internmc.facebook.com/intern/diff/D19926698/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Avoid clone for sparse tensors during accumulation of grads."

engine.
engine.
engine.
engine.
engine.
engine.
engine.

This PR is an attempt to avoid clone for sparse tensors similar to how
we avoid clone for dense tensors currently.

As per my understanding even if the 'indices' and 'values' of a sparse tensor
are non-continguous, operations like 'add' are still supported. As a result,
the major change in this PR is to use create a shallow copy instead of clone()
for sparse tensors.

Differential Revision: [D19926698](https://our.internmc.facebook.com/intern/diff/D19926698/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Avoid clone for sparse tensors during accumulation of grads."

engine.
engine.
engine.
engine.
engine.
engine.
engine.
engine.

This PR is an attempt to avoid clone for sparse tensors similar to how
we avoid clone for dense tensors currently.

As per my understanding even if the 'indices' and 'values' of a sparse tensor
are non-continguous, operations like 'add' are still supported. As a result,
the major change in this PR is to use create a shallow copy instead of clone()
for sparse tensors.

Differential Revision: [D19926698](https://our.internmc.facebook.com/intern/diff/D19926698/)

[ghstack-poisoned]

gqchen commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
```diff
@@ -1898,3 +1898,80 @@ def backward(ctx, grad):
+            for i in range(10):
+                dist_autograd.backward([loss], retain_graph=True)
+
+        # non-contiguous indices and value, we should trigger a copy.
```
do we want to test contiguous indices and values?

gqchen reviewed (commented):
one comment on the test. Otherwise it looks good to me. 

@albanD, can you elaborate on why we think this might be BC breaking? The sparse new grad by definition is not referenced anywhere else.

albanD commented:
The BC-breaking change is about the state of the `_values` and `_indices` Tensors inside the SparseTensor. In particular, since the old code always use `.clone()`, these Tensors are always "contiguous". This is particularly important because if someone tries to modify these inplace, the behavior can be unexpected.

gqchen commented on torch/csrc/autograd/functions/accumulate_grad.h:
```diff
@@ -55,6 +55,24 @@ struct TORCH_API AccumulateGrad : public Node {
+      } else if (
+          !GradMode::is_enabled() && new_grad_copy.is_sparse() &&
+          new_grad_copy._indices().is_contiguous() &&
+          new_grad_copy._values().is_contiguous() &&
```
@albanD Maybe I missed something obvious. We do shallow copy only when indices and values are contiguous?

albanD replied:
Yes. If they are already contiguous and not used anywhere else, it is fine to reuse the same memory.
Otherwise, we use clone to get brand new memory.

gqchen reviewed (commented):
Maybe I missed something obvious. In this version of the code, we do shallow copy only when _values and _indices are contiguous?

albanD commented on torch/csrc/autograd/functions/accumulate_grad.h:
```diff
@@ -55,6 +55,24 @@ struct TORCH_API AccumulateGrad : public Node {
+          !GradMode::is_enabled() && new_grad_copy.is_sparse() &&
+          new_grad_copy._indices().is_contiguous() &&
+          new_grad_copy._values().is_contiguous() &&
+          new_grad_use_count <= num_expected_refs) {
```
Actually following the comment above. We want to check the use_count for the indices and values as well here no?

pritamdamania87 commented on torch/csrc/autograd/functions/accumulate_grad.h:
```diff
@@ -55,6 +55,24 @@ struct TORCH_API AccumulateGrad : public Node {
         update_grad(new_grad_copy.detach());
+      } else if (
+          !GradMode::is_enabled() && new_grad_copy.is_sparse() &&
+          new_grad_copy._indices().is_contiguous() &&
```
We need to add a check to ensure sparse tensors are always created with continguous indicies and values: https://github.com/pytorch/pytorch/issues/12633. I'd like to keep the checks we have here and fix sparse tensor constructor and the logic here in a followup PR.

pritamdamania87 replied:
@ezyang Was looking into some of the Sparse Tensor math and it seems like some operations do handle non-contiguous indices/values appropriately: https://bddppq.github.io/codebrowser/pytorch/pytorch/aten/src/ATen/native/sparse/SparseTensorMath.cpp.html#_ZN2at6native18add_out_sparse_cpuERNS_6TensorERKS1_S4_N3c106ScalarE. I'm wondering if its actually fine to build SparseTensors with non-contiguous indices/values?

ezyang replied:
Even if it is fine, I assure you that it is (1) not tested and (2) doesn't work on a bunch of kernels. And unlike noncontiguous dense tensors, where there is a clear value add for non-contiguity (views), there is not really any clear value add to handling these cases correctly. For example, in CUDA:

```
  if (is_same_tensor(r, dense_buffer)) {
    TORCH_CHECK(r_.is_contiguous(), "add: CUDA dense-sparse addition with a non-contiguous output tensor does not work; shout if you need it (see https://github.com/pytorch/pytorch/issues/1521 )");
```

ezyang replied:
It might be possible to "best effort" non-contiguous values/indices by sprinkling contiguous everywhere in sparse but it's not clear to me that this is actually worth it. cc'ing @kurtamohler @nikitaved @hameerabbasi  for any other thoughts.

hameerabbasi replied:
If we're using external libraries in PyTorch, like MKL, those basically demand that we have contiguous values/indices. All literature I've read also implies this indirectly.

Although, of course, we could use non-contiguous versions, that'd be a huge performance hit over the already significant algorithmic hit from using sparse arrays of the same `nnz` as dense arrays. I might be wrong here, but it'd also be bad for GPU sparse arrays.

So, yes, I agree with the sentiment that ensuring contiguousness is a best practice for sparse arrays.

albanD commented on torch/csrc/autograd/functions/accumulate_grad.h:
```diff
@@ -55,6 +55,24 @@ struct TORCH_API AccumulateGrad : public Node {
         update_grad(new_grad_copy.detach());
+      } else if (
+          !GradMode::is_enabled() && new_grad_copy.is_sparse() &&
+          new_grad_copy._indices().is_contiguous() &&
```
This should be a different PR I think.
So that it is easier for us to measure the impact of doing this and add the corresponding tests.

pritamdamania87 replied:
I see, is there an existing gh issue for this or should I create one?

pritamdamania committed:
Update on "Avoid clone for sparse tensors during accumulation of grads."

engine.
engine.
engine.
engine.
engine.
engine.
engine.
engine.
engine.

This PR is an attempt to avoid clone for sparse tensors similar to how
we avoid clone for dense tensors currently.

As per my understanding even if the 'indices' and 'values' of a sparse tensor
are non-continguous, operations like 'add' are still supported. As a result,
the major change in this PR is to use create a shallow copy instead of clone()
for sparse tensors.

Differential Revision: [D19926698](https://our.internmc.facebook.com/intern/diff/D19926698/)

[ghstack-poisoned]

ezyang commented on torch/csrc/autograd/functions/accumulate_grad.h:
```diff
@@ -50,6 +50,26 @@ struct TORCH_API AccumulateGrad : public Node {
+        // We only skip clone if indices and values themselves are contiguous
+        // for backward compatiblity reasons. Since without this optimization,
+        // earlier we would clone the entire SparseTensor which cloned indices
+        // and values.
```
ISTR, weren't you going to file a GitHub issue for the `detach()` issue and link it here?

ezyang commented on torch/testing/_internal/distributed/rpc/dist_autograd_test.py:
```diff
@@ -1932,3 +1932,80 @@ def backward(ctx, grad):
             self.assertTrue(p_a == p_g or p_b == p_g)
+
+    @dist_init
+    def test_no_grad_copy_sparse(self):
```
It's not so nice that the test has to be copy pasted like this. It would be good if at least the two versions of the test are cross-linked from each other.

pritamdamania87 replied:
I originally thought about doing this, but the distributed autograd tests are different enough (with the context, a different API and the need to use get_gradients for the grads map) that I wasn't sure how much we would be able to dedup effectively.

Additionally since we removed the optimization of the same grad tensor being used in different variables, the behavior between distributed autograd and local autograd for the same test is slightly different.

albanD commented on torch/csrc/autograd/functions/accumulate_grad.h:
```diff
@@ -50,6 +50,26 @@ struct TORCH_API AccumulateGrad : public Node {
+          !GradMode::is_enabled() && new_grad.is_sparse() &&
+          new_grad._indices().is_contiguous() &&
+          new_grad._values().is_contiguous() &&
+          new_grad._indices().use_count() <= num_expected_refs &&
```
Is that correct? Shouldn't `num_expected_refs` be 1 here?

albanD replied:
These will have extra refs to the sparse Tensors. But the indices and values Tensors should only be referenced by the sparse Tensor.

pritamdamania87 replied:
Yes, you are right. I'll fix this in a follow up PR since this one is already landing.

gchanan replied:
how do you know indices and values only have 1 refcount?  What if you have a pre-existing reference to the indices tensor, from before you created the sparse tensor?

gchanan replied:
ah ok, *checking* 1 should be okay.

pritamdamania committed:
Update on "Avoid clone for sparse tensors during accumulation of grads."

engine.
engine.
engine.
engine.
engine.
engine.
engine.
engine.
engine.
engine.

This PR is an attempt to avoid clone for sparse tensors similar to how
we avoid clone for dense tensors currently.

As per my understanding even if the 'indices' and 'values' of a sparse tensor
are non-continguous, operations like 'add' are still supported. As a result,
the major change in this PR is to use create a shallow copy instead of clone()
for sparse tensors.

Differential Revision: [D19926698](https://our.internmc.facebook.com/intern/diff/D19926698/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Avoid clone for sparse tensors during accumulation of grads."

engine.
engine.
engine.
engine.
engine.
engine.
engine.
engine.
engine.
engine.
engine.

This PR is an attempt to avoid clone for sparse tensors similar to how
we avoid clone for dense tensors currently.

As per my understanding even if the 'indices' and 'values' of a sparse tensor
are non-continguous, operations like 'add' are still supported. As a result,
the major change in this PR is to use create a shallow copy instead of clone()
for sparse tensors.

Differential Revision: [D19926698](https://our.internmc.facebook.com/intern/diff/D19926698/)

[ghstack-poisoned]

gchanan commented on test/test_autograd.py:
```diff
@@ -3235,6 +3235,79 @@ def backward(ctx, grad):
+
+        # Run backwards multiple times to ensure accumulation works.
+        for i in range(10):
+            loss.backward(retain_graph=True)
```
shouldn't we check the accumulation?  This should be easy because the grad_input is the same as the grad_output, right?
